### ISSUE TITLE
cannot migrate with a custom schema

### ISSUE THREAD

**Issue by krichards** [romeerez/orchid-orm#274]
We are using orchid at haven ;) - and are trying to upgrade to the latest version but bumping into a problem where we are using a custom schema. It used to work but seems to have broken.

I believe what's happening is that the Db adapter with the "schema" set is now being wrapped by a transactional adapter, and there "schema" is not being replicated from the wrapped adapter, so when migration runs - there does not appear to be a schema in createMigrationsTable.  If I modify this to use

 const { schema } = db.adapter;

This it appears to create the schema but then fails in getMigratedVersionsMap.

**Comment by romeerez** [comment]
Are you writing migrations manually (then it's something with a migration) or using a recently added migration generator that generates migrations from tables defined in the code? In the case of the latter, it might be a bug in the generator logic.

**Comment by krichards** [comment]
I tried the schema=oidc on the URL parameter and setting schema on the dbconfig with no success. 

When I added some debugging it seemed, the schema was configured on the adapter but a transactional adapter was wrapping it and so schema was undefined.

When I looked into the connection string option I could not find any references to using ?schema with the node postgres library, so not sure how that could work. Guessing I'm missing something.

We are not creating the schema in the migration script, are just using the rakedb library to do it. Using an older revision it worked, so I was wondering if any changes could explain it.

**Comment by romeerez** [comment]
I looked at the code, and I can see that no schema prop on TransactionAdapter is normal and expected.

I can't recall any particular changes to that area since last Autumn, though it's possible that an unintended breaking change could have a place.

What is the error you're getting?

**Comment by krichards** [comment]
The function createMigrationsTable checks for schema on the supplied db.

Then creates it if it doesn't exist. But I have a transactional adapter and so no schema and so no schema is created. 

Unless that's a red herring and the schema is created for you elsewhere?

When you are creating the database are you creating the schema as well? 

I'll dig out the error message in a bit, will need to boot my machine.

**Comment by krichards** [comment]
error: no schema has been selected to create in
    at /home/kevin/projects/haven/haven-identity/node_modules/.pnpm/pg@8.11.5/node_modules/pg/lib/client.js:526:17
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async TransactionAdapter.query (file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/pqb@0.31.2/node_modules/pqb/dist/index.mjs:5369:12)
    at async createMigrationsTable (file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/rake-db@2.20.14/node_modules/rake-db/dist/index.mjs:2893:5)
    at async file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/rake-db@2.20.14/node_modules/rake-db/dist/index.mjs:2948:13
    at async Adapter.transaction (file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/pqb@0.31.2/node_modules/pqb/dist/index.mjs:5305:18)
    at async file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/rake-db@2.20.14/node_modules/rake-db/dist/index.mjs:2946:11
    at async runCommand (file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/rake-db@2.20.14/node_modules/rake-db/dist/index.mjs:5644:5)
    at async migrate (file:///home/kevin/projects/haven/haven-identity/adapter/oidc/orchid-pgsql/dist/db/db-migrate.js:29:9)
    at async main (file:///home/kevin/projects/haven/haven-identity/adapter/oidc/orchid-pgsql/dist/task/main.js:45:5)

**Comment by romeerez** [comment]
The error is about lack of privileges: https://dba.stackexchange.com/a/106067

The user from the connection string cannot create a table on that schema because has no such right to create tables in that schema.

I'm looking closer into this, possibly there is a bug in creating schema, perhaps an error message could be improved.

**Comment by romeerez** [comment]
> Unless that's a red herring and the schema is created for you elsewhere?

`createMigrationsTable` indeed creates a schema and a table.

> When you are creating the database are you creating the schema as well?

Yes, it creates a custom schema, and a migrations table.

**Comment by romeerez** [comment]
It all works well if the db user is a superuser, also it's fine if the user is an owner of the database.

But if you first create a database, and then create a non-privileged user, then problems begin, and you need to grant all needed permissions to the user.

```sql
-- enables creating new tables and other objects, and accessing them
GRANT create, usage ON SCHEMA "schema-name" TO "db-user";
```

The table for keeping track of migrated versions may be already created previously, and migration going to fail with "permission denied for table schemaMigrations".

Then you should grant all abilities on all tables in the schema to the db user:

```sql
GRANT SELECT, INSERT, UPDATE, DELETE ON ALL TABLES IN SCHEMA "schema-name" TO "db-user";
```

After doing that, I still have an error locally for one particular migration in which the user lacks certain permits, so you shall grant more and more permits to the user as required by error messages. If it's not a big security deal, making the user to be an owner of the database would solve all this.

**Comment by krichards** [comment]
I'll take a look. Can't recall if the user is granted ownership of the db or not. But this used to work with rakedb 2.10.? I think it was.

**Comment by romeerez** [comment]
To me, it doesn't seem to be related to rake-db changes. Of course, there could be various changes over the last few months and some of them could be somehow related to schema or connection, but the error doesn't look like a bug in the lib, so far.

**Comment by krichards** [comment]
I know there was a change where you went from using withSchema (or something similar) in the function that I was looking at.

I updated through a few versions, but had to change the type definitions in order to get it to build, and at some point the migration stopped working but I was concentrating on getting it to build so don't know when the problem started.

Will try and tie it down later

**Comment by romeerez** [comment]
2.10.0 was published a year ago and there were ~130 updates since then. It's technically possible to pinpoint the breaking version on your side, but, even if we know the exact version where it's broken, I still need to know a way to reproduce the issue, so I can add an additional test to support this case.

I reproduced the exact same error "no schema has been selected to create in" by having a db user with missing permissions, it's not a bug so this couldn't be broken by update. If you didn't change anything outside of rake-db, so that the database, db user, and schema are created in exact same way as before, that means a bug in the lib. Are you sure that wasn't changed?

If you can rollback to the last working version and confirm that it works with the same db settings, it would confirm a bug.

**Comment by romeerez** [comment]
Just remembered, you provided a code with reproduction last time, and rake-db was configured in a highly custom way, being wrapped and abstracted, so it has a bigger potential for things to go wrong when updating.

**Comment by krichards** [comment]
I've distilled the code down to the bare bones, no migrations, no code to generate the config.

[standalone.zip](https://github.com/user-attachments/files/15532173/standalone.zip)

```
kevin@kevin-desktop:~/projects/haven/haven-identity/adapter/oidc-standalone$ npm run db up

> db
> tsx db/dbScript.ts up

/home/kevin/projects/haven/haven-identity/adapter/oidc-standalone/node_modules/pg/lib/client.js:526
          Error.captureStackTrace(err)
                ^

error: no schema has been selected to create in
    at /home/kevin/projects/haven/haven-identity/adapter/oidc-standalone/node_modules/pg/lib/client.js:526:17
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at TransactionAdapter.query (/home/kevin/projects/haven/haven-identity/adapter/oidc-standalone/node_modules/pqb/src/adapter.ts:238:12)
    at createMigrationsTable (/home/kevin/projects/haven/haven-identity/adapter/oidc-standalone/node_modules/rake-db/src/migration/migrationsTable.ts:22:5)
    at <anonymous> (/home/kevin/projects/haven/haven-identity/adapter/oidc-standalone/node_modules/rake-db/src/commands/migrateOrRollback.ts:114:19)
    at Adapter.transaction (/home/kevin/projects/haven/haven-identity/adapter/oidc-standalone/node_modules/pqb/src/adapter.ts:141:18)
    at <anonymous> (/home/kevin/projects/haven/haven-identity/adapter/oidc-standalone/node_modules/rake-db/src/commands/migrateOrRollback.ts:111:17)
    at runCommand (/home/kevin/projects/haven/haven-identity/adapter/oidc-standalone/node_modules/rake-db/src/rakeDb.ts:152:11) {

```

I think the problem is that when you use your tool to create a database it creates the **schema and versionsTable** using an adapter that is not transactional.  Whislt we are not using your tool to create the database, its a docker image that initialises the database with an owner.

I suspect if you drop your schema before running migrate you might get the same results as I'm seeing.

**Comment by romeerez** [comment]
> I think the problem is that when you use your tool to create a database it creates the schema and versionsTable using an adapter that is not transactional. Whislt we are not using your tool to create the database

But how the problem can be at the step you're not executing?

> its a docker image that initialises the database with an owner.

The error cause may be at this step, can you share the docker code that creates a database and owner?

**Comment by krichards** [comment]
It's a problem for us because we are  trying to migrate a database that doesn't have the schema, so when we run the dB migrate it doesn't have the schema, and can't create one because there is no schema defined on the transactional adapter, when it tries to create the versions table it can't because the  custom schema doesn't exist.

I'll manually create the schema tomorrow on the empty database and then try the upgrade.

**Comment by krichards** [comment]
As I expected ... if I create the schema manually after the docker database has started, then the pnpm db up works,  if I remove the schema then you get the error.  So running just db up/migrate on a database without the custom schema does not create the schema.

We could change the code to create the schema ourselves, but it would be good if the migrate creates the schema, if only to prevent other people getting the same error.

**Comment by romeerez** [comment]
I'm still not sure how exactly you're creating the database, user, and schema.
Can you share that part of the docker script?

**Comment by krichards** [comment]
The docker setup is codified, not just a dockerfile, so it's not easy to share as  I would need to share the docker library we use and then the project just for starting the db. However if you drop the schema so all you have is the database and an owner, then run the db migrate, and you will get the same results as  we see.

**Comment by krichards** [comment]
I've added a docker file with just an empty postgres db,  - you just need to change the db url that's hardcoded in the dbScript.ts file to your IP address.

[standalonev2.zip](https://github.com/user-attachments/files/15580439/standalonev2.zip)

**Comment by romeerez** [comment]
> However if you drop the schema so all you have is the database and an owner, then run the db migrate, and you will get the same results as we see.

Thank, that's it, finally I can reproduce it!
Will fix it shortly.

**Comment by krichards** [comment]
@romeerez once we get past this - we've  found an issue with the jsonText when using a custom schema.  I can raise another ticket for that if you like.

```
    await db.createTable('device_code', (t) => ({
      id: t.varchar(MAX_ID_LENGTH).primaryKey().index(),
      userCode: t.varchar(MAX_USERCODE_LENGTH).index(),
      grantId: t.varchar(MAX_ID_LENGTH).index(),
      data: t.jsonText(),
      expiresAt: t.timestamp(),
      consumedAt: t.timestamp().nullable(),
    }));
   ```

We get the error

```
2024-06-05 08:32:12 ERROR ADAPTER | Failed to migrate database
TypeError: schema.stringSchema is not a function
    at new JSONTextColumn (file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/pqb@0.31.2/node_modules/pqb/dist/index.mjs:1749:26)
    at Object.jsonText (file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/pqb@0.31.2/node_modules/pqb/dist/index.mjs:5131:14)
    at file:///home/kevin/projects/haven/haven-identity/adapter/oidc/orchid-pgsql/dist/db/migrations/20230810000000_createOidcTables.js:13:17
    at getColumnTypes (file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/pqb@0.31.2/node_modules/pqb/dist/index.mjs:5030:10)
    at createTable (file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/rake-db@2.20.14/node_modules/rake-db/dist/index.mjs:439:13)
    at Function.createTable (file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/rake-db@2.20.14/node_modules/rake-db/dist/index.mjs:1313:12)
    at Array.<anonymous> (file:///home/kevin/projects/haven/haven-identity/adapter/oidc/orchid-pgsql/dist/db/migrations/20230810000000_createOidcTables.js:9:14)
    at runMigration (file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/rake-db@2.20.14/node_modules/rake-db/dist/index.mjs:3129:23)
    at async migrateOrRollback (file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/rake-db@2.20.14/node_modules/rake-db/dist/index.mjs:3064:5)
    at async file:///home/kevin/projects/haven/haven-identity/node_modules/.pnpm/rake-db@2.20.14/node_modules/rake-db/dist/index.mjs:2955:26
```

**Comment by romeerez** [comment]
You were absolutely right from the beginning pointing to the transactional adapter, I just didn't know what exactly is wrong and I could see it only when reproduced.

I fixed the schema problem and also the `schema.stringSchema is not a function` error, publishing update.

It will be published in a couple of minutes, then update `rake-db` and confirm if it works well for you.

**Comment by romeerez** [comment]
The fix (for both) is now published

**Comment by krichards** [comment]
Great. Thanks for your help @romeerez!

**Comment by romeerez** [comment]
I'm glad to know it's working for you, feel free to post more issues if anything.

As for the past issue (from Autumn), I believe that it was fixed because:
- `db up` creates a lock in Postgres.
- when running `db up` in parallel (by a separate pod or process) it will wait for the first to finish.
- after migrations are done, the lock is released, and parallel processes will only find out that there is nothing to migrate.

But race conditions sound alarming and they must be fixed with a top priority, so if you see any symptoms I'll investigate it in the first order.

**Comment by krichards** [comment]
> @romeerez I'm an engineer working with @krichards and having updated the the latest version I can see the schema is now created and our migrations start to run so thank for fixing this  👍. 
> 
> We do have a separate issue around a race condition when running the migrations, that I think we've raised before.
> 
> But we're still looking into that and will raise a separate issue if we feel we need to.

I was wondering whether the latest version might fix the migration running. It seemed quite repeatable with the latest version. I was wondering whether it was related to checking if the version table existed combined with  transactionable migrations, an exception cause the whole thing to rollback whereas before I think it would carry on.

Will leave it with you jon. Cheers
